sat/apiclient/pcs.py

```diff
@@ -0,0 +1,192 @@
+
+            return f'{self.message}\n{xname_err_summary}'
+
+
+class PCSClient(APIGatewayClient):
+    """Client for the Power Control Service."""
+    base_resource_path = 'power-control/v1/'
+
+    def set_xnames_power_state(self, xnames, power_state, force=False, recursive=False, prereq=False):
+        """Set the power state of the given xnames.
+
+        Args:
+            xnames (list): the xnames (str) to perform the power operation
+                against.
+            power_state (str): the desired power state. Either "on" or "off".
+            force (bool): if True, disable checks and force the power operation.
+            recursive (bool): if True, power on component and its descendants.
+            prereq (bool): if True, power on component and its ancestors.
+
+        Returns:
+            None
+
+        Raises:
+            ValueError: if the given `power_state` is not one of 'on' or 'off'
+            PCSError: if the attempt to power on/off the given xnames with PCS
+                fails. This exception contains more specific information about
+                the failure, which will be included in its __str__.
+        """
+        allowed_states = {'on', 'off', 'soft-off', 'soft-restart', 'hard-restart', 'init', 'force-off'}
+        power_state = power_state.lower()
+        if power_state not in allowed_states:
+            allowed_states_str = ", ".join("\"" + state + "\"" for state in allowed_states)
+            raise ValueError(f'Invalid power state {power_state} given. Must be {allowed_states_str}')
+
+        if force:
+            if power_state in {'off', 'soft-off'}:
+                power_state = 'force-off'
+            elif power_state == 'soft-restart':
+                power_state = 'hard-restart'
+
+        target_xnames = set()
+        if recursive:
+            hsm_client = HSMClient(self.session)
+
+            try:
+                target_xnames |= set(
+                    component['XName']
+                    for xname in xnames
+                    for component in hsm_client.get_node_components(ancestor=xname)
+                )
+            except APIError as err:
+                raise PCSError(f'Could not retrieve descendent components for xnames: {err}') from err
+        if prereq:
+            hsm_client = HSMClient(self.session)
+            try:
+                target_xnames |= set(
+                    component['XName']
+                    for component in hsm_client.get_all_components()
+                    if any(xname.startswith(component['XName']) for xname in xnames)
+                )
+            except APIError as err:
+                raise PCSError(f'Could not query ancestor components for xnames: {err}') from err
+
+        params = {
+            'operation': power_state,
+            'taskDeadlineMinutes': -1,
+            'location': [
+                {'xname': xname}
+                for xname in target_xnames
```

Comment:
@shivaprasad-metimath, I think I've found the cause of the problem you encountered in your testing on odin. This `target_xnames` value is defined as an empty set on line 106. It has components added to it on lines 111 and 121, but only if `recursive` and `prereq` are defined. It never adds the xnames which were actually passed in as the value of the `xnames` argument to this method.